internal/keys/keys.go

```diff
@@ -0,0 +1,158 @@
+	"k8s.io/client-go/util/keyutil"
+)
+
+const duration365d = time.Hour * 24 * 365
+
+// SignerWithPEM is a signer implementation that has its PEM encoded representation attached
+type SignerWithPEM interface {
+	crypto.Signer
+	PEM() []byte
+}
+
+type signerWithPEM struct {
+	crypto.Signer
+	pemData []byte
+}
+
+func (s signerWithPEM) PEM() []byte {
+	return s.pemData
+}
+
+// CA is a certificate authority usable to generated signed certificates
+type CA interface {
+	PrivateKey() SignerWithPEM
+	Cert() *x509.Certificate
+	NewSignedCert(cfg certutil.Config, publicKey crypto.PublicKey) (*x509.Certificate, error)
+}
+
+type authority struct {
+	pvk  SignerWithPEM
+	cert *x509.Certificate
+}
+
+func (ca *authority) PrivateKey() SignerWithPEM {
+	return ca.pvk
+}
+
+func (ca *authority) Cert() *x509.Certificate {
+	return ca.cert
+}
+
+func (ca *authority) NewSignedCert(cfg certutil.Config, publicKey crypto.PublicKey) (*x509.Certificate, error) {
+	serial, err := cryptorand.Int(cryptorand.Reader, new(big.Int).SetInt64(math.MaxInt64))
+	if err != nil {
+		return nil, err
+	}
+	if len(cfg.CommonName) == 0 {
+		return nil, errors.New("must specify a CommonName")
+	}
+	if len(cfg.Usages) == 0 {
+		return nil, errors.New("must specify at least one ExtKeyUsage")
+	}
+	certTmpl := x509.Certificate{
+		Subject: pkix.Name{
+			CommonName:   cfg.CommonName,
+			Organization: cfg.Organization,
+		},
+		DNSNames:     cfg.AltNames.DNSNames,
+		IPAddresses:  cfg.AltNames.IPs,
+		SerialNumber: serial,
+		NotBefore:    ca.cert.NotBefore,
+		NotAfter:     time.Now().Add(time.Hour * 24 * 365).UTC(),
+		KeyUsage:     x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:  cfg.Usages,
+	}
+	certDERBytes, err := x509.CreateCertificate(cryptorand.Reader, &certTmpl, ca.cert, publicKey, ca.pvk)
+	if err != nil {
+		return nil, err
+	}
+	cert, err := x509.ParseCertificate(certDERBytes)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+// NewSelfSignedCA creates a CA and return it with its private key
+func NewSelfSignedCA(commonName string, organization []string) (CA, error) {
+	pvk, err := ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	derBytes, err := x509.MarshalECPrivateKey(pvk)
+	if err != nil {
+		return nil, err
+	}
+	keyPemBlock := &pem.Block{
+		Type:  keyutil.ECPrivateKeyBlockType,
+		Bytes: derBytes,
+	}
+	keyData := pem.EncodeToMemory(keyPemBlock)
+	key := signerWithPEM{Signer: pvk, pemData: keyData}
+
+	now := time.Now()
+	tmpl := x509.Certificate{
+		SerialNumber: new(big.Int).SetInt64(0),
+		Subject: pkix.Name{
+			CommonName:   commonName,
+			Organization: organization,
+		},
+		NotBefore:             now.UTC(),
+		NotAfter:              now.Add(duration365d * 10).UTC(),
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
```

Comment:
No SubjectKeyId? It's not mandatory but it's a good practice

Reply:
I am a bit worried about modifying this. I am not an expert in x509, and this is actually extracted from pre-1.14 kubernetes/client-go tools to generate certificates (with a refactoring making it fit better the needs of Compose on Kubernetes), wich we are pretty confident that it is working for our use-case.

What is exactly the value of SubjectKeyId and what advantage would it bring for our use case ?